packages/data-point/lib/entity-types/base-entity/resolve.js

```diff
@@ -127,52 +139,87 @@ function resolveEntity (
     currentAccumulator.trace === true ||
     currentAccumulator.reducer.spec.params.trace === true
 
-  let accUid = currentAccumulator
   let timeId
+  let accUid = currentAccumulator
   if (trace === true) {
     accUid = utils.set(currentAccumulator, 'euid', utils.getUID())
     timeId = `⧖ ${accUid.context.id}(${accUid.euid})`
     console.time(timeId)
   }
 
-  const resolveReducerBound = resolveReducer.bind(null, manager)
+  const {
+    inputType,
+    before,
+    after,
+    outputType
+  } = currentAccumulator.reducer.spec
 
-  return Promise.resolve(accUid)
-    .then(acc =>
-      typeCheck(manager, acc, acc.reducer.spec.inputType, resolveReducer)
-    )
-    .then(acc => resolveMiddleware(manager, `before`, acc))
-    .then(acc =>
-      resolveMiddleware(manager, `${reducer.entityType}:before`, acc)
-    )
-    .then(acc => resolveReducer(manager, acc, acc.reducer.spec.before))
-    .then(acc => mainResolver(acc, resolveReducerBound))
-    .then(acc => resolveReducer(manager, acc, acc.reducer.spec.after))
-    .then(acc => resolveMiddleware(manager, `${reducer.entityType}:after`, acc))
-    .then(acc => resolveMiddleware(manager, `after`, acc))
-    .catch(error => {
-      // checking if this is an error to bypass the `then` chain
-      if (error.bypass === true) {
-        return error.bypassValue
-      }
+  let result = Promise.resolve(accUid)
 
-      throw error
+  if (inputType) {
+    result = result.then(acc => {
+      return typeCheck(manager, acc, inputType, resolveReducer)
     })
-    .then(acc =>
-      typeCheck(manager, acc, acc.reducer.spec.outputType, resolveReducer)
-    )
+  }
+
+  result = resolveMiddleware(manager, result, 'before')
+
+  result = resolveMiddleware(manager, result, `${reducer.entityType}:before`)
+
+  if (before) {
+    result = result.then(acc => {
+      return resolveReducer(manager, acc, before)
+    })
+  }
+
+  result = result.then(acc => {
+    return mainResolver(acc, resolveReducer.bind(null, manager))
+  })
+
+  if (after) {
+    result = result.then(acc => {
+      return resolveReducer(manager, acc, after)
+    })
+  }
+
+  result = resolveMiddleware(manager, result, `${reducer.entityType}:after`)
+
+  result = resolveMiddleware(manager, result, 'after')
+
+  result = result.catch(error => {
+    // checking if this is an error to bypass the `then` chain
+    if (error.bypass === true) {
+      return error.bypassValue
+    }
+
+    throw error
+  })
+
+  if (outputType) {
+    result = result.then(acc => {
+      return typeCheck(manager, acc, outputType, resolveReducer)
+    })
+  }
+
+  return result
     .catch(error => {
       // attach entity information to help debug
       error.entityId = currentAccumulator.reducer.spec.id
 
-      return resolveErrorReducers(
+      let result = resolveErrorReducers(
```

Comment:
can you change the name of this variable to avoid naming scope conflicts while a human reads it

Reply:
done